[SeperateConstOffsetFromGEP] Handle `or disjoint` flags

krzysz00 opened the pull request:
This commit extends separate-const-offset-from-gep to look at the newly-added `disjoint` flag on `or` instructions so as to preserve additional opportunities for optimization.

The tests were pre-committed in #76972.

krzysz00 committed:
[SeperateConstOffsetFromGEP] Handle `or disjoint` flags

This commit extends separate-const-offset-from-gep to look at the
newly-added `disjoint` flag on `or` instructions so as to preserve
addditional opportunities for optimization.

As with other `or disjoint`-handling commits, this does not remove the
existing check for the or's operands having no bits in common because
`disjoint` is currently not inferred.

The tests were pre-committed in #76972.

nikic commented:
> As with other `or disjoint`-handling commits, this does not remove the existing check for the or's operands having no bits in common because `disjoint` is currently not inferred.

disjoint is inferred on the IR level, so you should remove the haveNoCommonBitsSet() call (and adjust tests that depend on it), unless there is some special circumstance here that requires otherwise.

krzysz00 commented:
Huh. I was reading commit logs that implied  that that wasn't the case yet. I can adjust

topperc commented:
> Huh. I was reading commit logs that implied that that wasn't the case yet. I can adjust

It's not inferred in SelectionDAG yet. Maybe you read the log from one of my SelectionDAG commits?

krzysz00 commented:
Looks like removing the NoBitsInCommon checks breaks an existing test - would it make more sense to go annotate said test with `disjoint` or to keep the check?

nikic commented:
> Looks like removing the NoBitsInCommon checks breaks an existing test - would it make more sense to go annotate said test with `disjoint` or to keep the check?

Annotate the test, assuming the flag would be present coming out of the middle-end (presumably yes).

krzysz00 commented:
Update: having actually run more tests, removing the bitsInCommon thing broke a bunch of AMD codegen tests and might break some other platforms I don't usually build.

Do we know if there's a way to go run `disjoint` inference on existing IR?

nikic commented:
I don't think there is an easy way to *only* add disjoint. I just did this manually where necessary (e.g. eecb99c5f66c8491766628a2925587e20f3b1dbd adds it to about a hundred tests).

krzysz00 committed:
Merge branch 'main' into disjoint-or-seperate-const-gep

krzysz00 committed:
Remove old check, update tests.

All tests that were relying on haveNoBitsInCommon() checks have been
in SeparateConstOffsetFromGEP have been updated to have the relevant
`or`s annotated with `disjoint`.

arsenm commented:
Not sure if we need additional negative tests for missing disjoints
